examples/paillier_logistic_regression.py

```diff
@@ -0,0 +1,245 @@
+
+np.random.seed(42)
+
+# Enron spam dataset
+url1 = 'http://www.aueb.gr/users/ion/data/enron-spam/preprocessed/enron1.tar.gz'
+url2 = 'http://www.aueb.gr/users/ion/data/enron-spam/preprocessed/enron2.tar.gz'
+
+
+def download_data():
+    """Download two sets of Enron1 spam/ham emails if they are not here"""
+
+    if (not os.path.isdir('examples/enron1') or
+       not os.path.isdir('examples/enron2')):
+
+        print("Downloading 1/2:", url1)
+        # First folder -> train set
+        foldertar = 'examples/emails.tar.gz'
+
+        with urlopen(url1) as remotedata:
+            with open(foldertar, 'wb') as tar:
+                tar.write(remotedata.read())
+
+        with tarfile.open(foldertar) as tar:
+            tar.extractall('examples/')
+        os.remove(foldertar)
+
+        # Second folder -> test set
+        print("Downloading 2/2:", url2)
+        with urlopen(url2) as remotedata:
+            with open(foldertar, 'wb') as tar:
+                tar.write(remotedata.read())
+
+        with tarfile.open(foldertar) as tar:
+            tar.extractall('examples/')
+        os.remove(foldertar)
+
+
+def preprocess_data():
+    """
+    Get the Enron emails from disk.
+    Represent them as bag-of-words.
+    Shuffle and split train/test.
+    """
+
+    print("Importing dataset from disk...")
+    path = 'examples/enron1/ham/'
+    ham1 = [open(path + f, 'r', errors='replace').read().strip(r"\n")
+            for f in os.listdir(path) if os.path.isfile(path + f)]
+    path = 'examples/enron1/spam/'
+    spam1 = [open(path + f, 'r', errors='replace').read().strip(r"\n")
+             for f in os.listdir(path) if os.path.isfile(path + f)]
+    path = 'examples/enron2/ham/'
+    ham2 = [open(path + f, 'r', errors='replace').read().strip(r"\n")
+            for f in os.listdir(path) if os.path.isfile(path + f)]
+    path = 'examples/enron2/spam/'
+    spam2 = [open(path + f, 'r', errors='replace').read().strip(r"\n")
+             for f in os.listdir(path) if os.path.isfile(path + f)]
+
+    # Merge and create labels
+    X = ham1 + spam1 + ham2 + spam2
+    y = np.array([-1] * len(ham1) + [1] * len(spam1) +
+                 [-1] * len(ham2) + [1] * len(spam2))
+
+    # Words count, keep only fequent words
+    count_vect = CountVectorizer(decode_error='replace', stop_words='english',
+                                 min_df=0.001)
+    X = count_vect.fit_transform(X)
+
+    print('Vocabulary size: %d' % X.shape[1])
+
+    # Shuffle
+    perm = np.random.permutation(X.shape[0])
+    X, y = X[perm, :], y[perm]
+
+    # Split train and test
+    split = 500
+    X_train, X_test = X[-split:, :], X[:-split, :]
+    y_train, y_test = y[-split:], y[:-split]
+
+    print("Labels in trainset are %.2f spam / %.2f ham"
+          % (np.mean(y_train == 1), np.mean(y_train == -1)))
+
+    return X_train, y_train, X_test, y_test
+
+
+class TimeIt():
+    """Helper for measuring runtime"""
+
+    def tick(self):
+        self.time0 = time.time()
```

Comment:
I acknowledge that I'm just being pedantic here. Because we only care about the absolute difference the best option is [time.perf_counter](https://docs.python.org/3/library/time.html#time.perf_counter), the next best option is probably the monotonic counter, and then finally (but still very acceptable) `time.time()`